core/src/main/scala/org/apache/spark/Accumulators.scala

```diff
@@ -320,7 +328,12 @@ private[spark] object Accumulators {
   def add(values: Map[Long, Any]): Unit = synchronized {
     for ((id, value) <- values) {
       if (originals.contains(id)) {
-        originals(id).asInstanceOf[Accumulable[Any, Any]] ++= value
+        // Since we are now storing weak references, we must check whether the underlying data
+        // is valid. 
+        originals(id).get match {
+          case Some(accum) => accum.asInstanceOf[Accumulable[Any, Any]] ++= value
+          case None => // Do nothing   
```

Comment:
Should this be an error-case?  This seems to mean that you've received an accumulator update for an accumulator that's been garbage-collected, which seems like it should be an error.